pygfx/controllers/_input.py

```diff
@@ -0,0 +1,121 @@
+    def __init__(self, register_events=None):
+        # TODO:
+        # - controllers
+        # - virtual input state (e.g. buttons and axes, so users can remap controls)
+
+        self._key_down = {}
+        self._key_up = {}
+        self._key_state = {}
+
+        self._pointer_down = {}
+        self._pointer_up = {}
+        self._pointer_state = {}
+        self._pointer_pos = {}
+        self._pointer_pos_prev = {}
+        self._pointer_delta = {}
+
+        self._wheel_delta = {}
+
+        if register_events is not None:
+            self.register_events(register_events)
+
+    def register_events(self, viewport_or_renderer):
+        viewport = Viewport.from_viewport_or_renderer(viewport_or_renderer)
+        viewport.renderer.add_event_handler(
+            lambda event: self.handle_event(event, viewport),
+            "pointer_down",
+            "pointer_move",
+            "pointer_up",
+            "wheel",
+            "key_down",
+            "key_up",
+            "after_flush",
+        )
+
+    def handle_event(self, event, viewport):
+        type = event.type
+
+        if type == "after_flush":
+            self._pointer_pos_prev = self._pointer_pos.copy()
+            self._key_down.clear()
+            self._key_up.clear()
+            self._pointer_down.clear()
+            self._pointer_up.clear()
+            self._pointer_delta.clear()
+            self._wheel_delta.clear()
+
+        elif type.startswith("pointer_"):
+            pointer_id = event.pointer_id
+
+            if type == "pointer_move":
+                x, y = event.x, event.y
+                self._pointer_pos[pointer_id] = (x, y)
+
+                # prevent emitting a bogus delta when we don't have a previous
+                # position for the mouse
+                if (prev := self._pointer_pos_prev.get(pointer_id)) is not None:
+                    self._pointer_delta[pointer_id] = (
+                        x - prev[0],
+                        y - prev[1],
+                    )
+            elif type == "pointer_down":
+                key = (pointer_id, event.button)
+                self._pointer_state[key] = True
+                self._pointer_down[key] = True
+            elif type == "pointer_up":
+                key = (pointer_id, event.button)
+                self._pointer_state[key] = False
+                self._pointer_up[key] = True
+
+        elif type.startswith("key_"):
+            key = event.key.lower()
+
+            if type == "key_down":
+                self._key_state[key] = True
+                self._key_down[key] = True
+            elif type == "key_up":
+                self._key_state[key] = False
+                self._key_up[key] = True
+
+        elif type == "wheel":
+            self._wheel_delta[event.pointer_id] = event.dy
+
+    def key_down(self, key):
+        """Returns True if the key_down event was received in the current frame"""
+        return self._key_down.get(key.lower(), False)
```

Comment:
Is there a benefit to this, compared to listening for key_down events directly? Also in theory a user can press a key twice, and using this method only see one. Or is that intentional?

Reply:
It's an abstraction that makes it possible for many once-per-frame update functions to avoid the boilerplate of receiving events and tracking state of inputs :)

Reply:
I see the convenience, but it also add the constraint that any code that uses this must run "in sync", i.e. exactly once per draw.

Reply:
Yes, although that is intentional

Reply:
I also wanted to mention, in an earlier state of this code I had used `input.key(...)` instead of `input.key_down(...)`. It was very tricky to toggle fly mode because pressing the key would make `input.key(...)` evaluate to `True` for multiple frames. In other words, it would rapidly toggle where I just wanted to toggle it once. That's the use case for `input.key_down(...)`.